File: static/js/zabbix_api.js
function initEcharts(divId, name, time, key, historyType, limit, auth) {
    var url = 'http://192.168.12.243/zabbix/api_jsonrpc.php';
    // auth = '541374ddb8b6c66b9e911f0b2f8e0440';
    var myChart = echarts.init(document.getElementById(divId));
// 显示标题，图例和空的坐标轴
    myChart.setOption({
        title: {
            text: 'CPU',
            left: '50%',
            textAlign: 'center'
        },
        tooltip: {
            trigger: 'axis',
            axisPointer: {
                lineStyle: {
                    color: '#ddd'
                }
            },
            backgroundColor: 'rgba(255,255,255,1)',
            padding: [5, 10],
            textStyle: {
                color: '#7588E4'
            },
            extraCssText: 'box-shadow: 0 0 5px rgba(0,0,0,0.3)'
        },
        legend: {
            right: 20,
            orient: 'vertical',
            data: 'cpu'
        },
        xAxis: {
            type: 'category',
            data: [],
            boundaryGap: false,
            splitLine: {
                show: true,
                interval: 'auto',
                lineStyle: {
                    color: ['#D4DFF5']
                }
            },
            axisTick: {
                show: false
            },
            axisLine: {
                lineStyle: {
                    color: '#609ee9'
                }
            },
            axisLabel: {
                margin: 10,
                textStyle: {
                    fontSize: 14
                }
            }
        },
        yAxis: {
            type: 'value',
            splitLine: {
                lineStyle: {
                    color: ['#D4DFF5']
                }
            },
            axisTick: {
                show: false
            },
            axisLine: {
                lineStyle: {
                    color: '#609ee9'
                }
            },
            axisLabel: {
                margin: 10,
                textStyle: {
                    fontSize: 14
                }
            }
        },
        series: [{
            name: name,
            type: 'line',
            smooth: true,
            showSymbol: false,
            symbol: 'circle',
            areaStyle: {
                normal: {
                    color: new echarts.graphic.LinearGradient(0, 0, 0, 1, [{
                        offset: 0,
                        color: 'rgba(199, 237, 250,0.5)'
                    }], false)
                }
            },
            itemStyle: {
                normal: {
                    color: '#3598DC'
                }
            },

            data: []


        }]
    });
    myChart.showLoading();    //数据加载完之前先显示一段简单的loading动画
    setInterval(function () {
        // login(url, user, password, key, historyType, limit, myChart);
        hostsidGet(url, auth, 'kvm-master', key, historyType, limit, myChart);
    }, time);
}

// function login(url, user, password, key, historyType, limit, myChart) {
//     $.ajax({
//         type: "post",
//         async: true,  //异步请求（同步请求将会锁住浏览器，用户其他操作必须等待请求完成才可以执行）
//         url: url,
//         contentType: "application/json",
//         data: JSON.stringify({
//             "jsonrpc": "2.0",
//             "method": "user.login",
//             "id": 1,
//             "params": {
//                 "user": user,
//                 "password": password
//             }
//         }),
//         dataType: "json",        //返回数据形式为json
//         success: function (json) {
//             //请求成功时执行该函数内容，result即为服务器返回的json对象
//             var auth = json.result;
//             console.log(auth);
//
//             hostsidGet(url, auth, 'Zabbix server', key, historyType, limit, myChart);
//
//         },
//
//         error: function (errorMsg) {
//             //请求失败时执行该函数
//             console.log(errorMsg);
//         }
//     });
// }

function hostsidGet(url, auth, hostName, key, historyType, limit, myChart) {
    $.ajax({
        type: "post",
        async: true,  //异步请求（同步请求将会锁住浏览器，用户其他操作必须等待请求完成才可以执行）
        url: url,    //请求发送到TestServlet处
        contentType: "application/json",
        data: JSON.stringify({
            "jsonrpc": "2.0",
            "method": "host.get",
            "params": {
                "filter": {
                    "host": [hostName]
                },
                "output": [
                    "hostid",
                    "host"
                ]
            },
            "id": 2,
            "auth": auth
        }),
        dataType: "json",        //返回数据形式为json
        success: function (json) {
            //请求成功时执行该函数内容，result即为服务器返回的json对象
            var hosts = json.result;
            // console.log(json.result[0].hostid);
            console.log(hosts);
            for (var i = 0, len = hosts.length; i < len; i++) {
                console.log(hosts[i].hostid);
                itemsGet(url, auth, hosts[i].hostid, key, historyType, limit, myChart);
            }
        },

        error: function (errorMsg) {
            //请求失败时执行该函数
            console.log(errorMsg);
        }
    });
}

function itemsGet(url, auth, hostId, key, historyType, limit, myChart) {
    $.ajax({
        type: "post",
        async: true,  //异步请求（同步请求将会锁住浏览器，用户其他操作必须等待请求完成才可以执行）
        url: url,    //请求发送到TestServlet处
        contentType: "application/json",
        data: JSON.stringify({
            "jsonrpc": "2.0",
            "method": "item.get",
            "params": {
                "output": [
                    "key_",
                    "name",
                    'itemid'
                ],
                "hostids": hostId,
                "filter": {
                    "key_": key
                }
            },
            "auth": auth,
            "id": 3
        }),
        dataType: "json",        //返回数据形式为json
        success: function (json) {
            //请求成功时执行该函数内容，result即为服务器返回的json对象
            var items = json.result;
            console.log(items);

            for (var i = 0, len = items.length; i < len; i++) {
                console.log(items[i].itemid);
                historysGet(url, auth, items[i].itemid, historyType, limit, myChart);
            }
        },

        error: function (errorMsg) {
            //请求失败时执行该函数
            console.log(errorMsg);
        }
    });
}

function historysGet(url, auth, itemId, historyType, limit, myChart) {
    $.ajax({
        type: "post",
        async: true,  //异步请求（同步请求将会锁住浏览器，用户其他操作必须等待请求完成才可以执行）
        url: url,    //请求发送到TestServlet处
        contentType: "application/json",
        data: JSON.stringify({
            "jsonrpc": "2.0",
            "method": "history.get",
            "params": {
                "output": "extend",
                "history": historyType,
                "itemids": itemId,
                "sortfield": "clock",
                "sortorder": "DESC",
                "limit": limit
            },
            "auth": auth,
            "id": 4
        }),
        dataType: "json",        //返回数据形式为json
        success: function (json) {
            //请求成功时执行该函数内容，result即为服务器返回的json对象
            var historys = json.result;
            console.log(historys);

            var dates = [];
            var values = [];
            for (var i = 0, len = historys.length; i < len; i++) {
                dates.push(formatUnixtimestamp(historys[i].clock));
                values.push(historys[i].value);
            }
            dates.reverse();
            values.reverse();
            myChart.hideLoading();    //隐藏加载动画
            myChart.setOption({        //加载数据图表
                xAxis: {
                    data: dates
                },
                series: [{
                    // 根据名字对应到相应的系列
                    data: values
                }]
            });
        },

        error: function (errorMsg) {
            //请求失败时执行该函数
            console.log(errorMsg);
        }
    });
}

function formatUnixtimestamp(unixTimeStamp) {
    var timeStamp = new Date(unixTimeStamp * 1000);
    var year = 1900 + timeStamp.getYear();
    var month = "0" + (timeStamp.getMonth() + 1);
    var date = "0" + timeStamp.getDate();
    var hour = "0" + timeStamp.getHours();
    var minute = "0" + timeStamp.getMinutes();
    var second = "0" + timeStamp.getSeconds();
    return year + "-" + month.substring(month.length - 2, month.length) + "-" + date.substring(date.length - 2, date.length)
        + " " + hour.substring(hour.length - 2, hour.length) + ":"
        + minute.substring(minute.length - 2, minute.length) + ":"
        + second.substring(second.length - 2, second.length);
}

///////////////

File: res/zabbix_api.py
import requests

# url = 'http://192.168.8.129/zabbix/api_jsonrpc.php'


class Zabbix:
    def __init__(self, url, user, password):
        '''
        初试化zabbix对象
        :param url: str, 'http://ip/zabbix/api_jsonrpc.php'
        :param user: str, 'Admin'
        :param password: str, 'zabbix'
        '''
        self.url = url
        self.user = user
        self.password = password
        self.login()

    def login(self):
        ''' 登录并获取auth口令 '''
        data = {
            "jsonrpc": "2.0",
            "method": "user.login",
            "id": 1,
            "params": {
                "user": self.user,
                "password": self.password
            }
        }
        response = requests.post(url=self.url, json=data).json()

        if response.get('error'):
            print(response.get('error'))
        elif response.get('result'):
            self.auth = response.get('result')

        return response.get('result')

    def hostid_get(self, host_name):
        '''
        获取host_name的hostid用于获取其被监控的参数
        :param host_name: str, 'Zabbix server'
        :return: list, [{'hostid': '10084', 'host': 'Zabbix server'}]
        '''
        data = {
            "jsonrpc": "2.0",
            "method": "host.get",
            "params": {
                "filter": {
                    "host": [host_name]
                },
                "output": [
                    "hostid",
                    "host",
                    "available"
                ]
            },
            "id": 2,
            "auth": self.auth
        }
        response = requests.post(url=self.url, json=data).json()
        # print('hostid_get response:\n', response)

        return response.get('result', [])

    def itemid_get(self, hostid, key):
        '''
        根据hostid和key（监控项的键名）来获取监控项的id
        :param hostid: str, '10084'
        :param key: str, 'system.cpu.load[percpu,avg1]'
        :return:list, [{'key_': 'system.cpu.load[percpu,avg1]', 'itemid': '23296',
                        'name': 'Processor load (1 min average per core)' }]
        '''
        data = {
            "jsonrpc": "2.0",
            "method": "item.get",
            "params": {
                "output": [
                    "key_",
                    "name",
                    'itemid',
                    # "extend"
                ],
                "hostids": hostid,
                "filter": {
                    "key_": key
                },
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        # print('itemid_get response:\n', response)

        return response.get('result', [])

    def _historys_get(self, itemid, history_type, limit):
        '''
        根据监控项的id获取其历史记录
        :param itemid: str, '23296'
        :param history_type:0 - numeric float;
                            1 - character;
                            2 - log;
                            3 - numeric unsigned;
                            4 - text.
                            Default: 3.
        :param limit: int, 需要放回的记录的条数
        :return:
        '''
        data = {
            "jsonrpc": "2.0",
            "method": "history.get",
            "params": {
                "output": "extend",
                "history": history_type,
                "itemids": itemid,
                "sortfield": "clock",
                "sortorder": "DESC",
                "limit": limit
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        print('_historys_get response:\n', response)

        return response.get('result', [])

    def historys_get(self, host_name, key, history_type, limit):
        """
        根据四个参数获监控项的历史记录
        :param host_name:
        :param key:
        :param history_type:
        :param limit:
        :return:
        """
        hosts = self.hostid_get(host_name)
        print('hosts:\n', hosts)
        for host in hosts:
            if host["available"] == "1":
                items = self.itemid_get(host['hostid'], key)
                print('items:\n', items)
                for item in items:
                    historys = self._historys_get(item['itemid'], history_type, limit)
                    print('historys:\n', historys)
                    return historys
            else:
                return [
                    {
                        "value": -1
                    }
                ]

    def create_host(self, host_name, host_ip):
        """
        创建zabbix监控的主机,目前将模板定为Template OS Linux(id=10001)，组定为Linux servers(id=2)
        :param host_name: str, 主机名
        :param host_ip: str, 主机ip
        :return:
        """

        data = {
            "jsonrpc": "2.0",
            "method": "host.create",
            "params": {
                "host": host_name,
                "interfaces": [
                    {
                        "type": 1,
                        "main": 1,
                        "useip": 1,
                        "ip": host_ip,
                        "dns": "",
                        "port": 10050
                    }
                ],
                "groups": [
                    {
                        "groupid": "2"
                    }
                ],
                "templates": [
                    {
                        "templateid": "10001"
                    }
                ]
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        return response.get('result', {})

    def get_template_id(self, template_name):
        """
        根据模板名称来获取模板的id，将来可能需要为节点设置不同的模板
        :param template_name: 模板名称
        :return:
        """

        data = {
            "jsonrpc": "2.0",
            "method": "template.get",
            "params": {
                "output": "templateid",
                "filter": {
                    "host": template_name
                }
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        return response.get('result', [])

    def get_group_id(self, group_name):
        """
        根据组名来获取组id，将来可能需要为节点设置不同的组
        :param group_name: 组名
        :return:
        """

        data = {
            "jsonrpc": "2.0",
            "method": "hostgroup.get",
            "params": {
                "output": "groupid",
                "filter": {
                    "name": group_name
                }
            },
            "auth": self.auth,
            "id": 1
        }

        response = requests.post(url=self.url, json=data).json()
        return response.get('result', [])

    def get_all_hosts(self):
        data = {
            "jsonrpc": "2.0",
            "method": "host.get",
            "params": {
                "output": [
                    "hostid",
                    "host"
                ],
                "selectInterfaces": [
                    "interfaceid",
                    "ip",
                    "available"
                ]
            },
            "id": 1,
            "auth": self.auth
        }

        response = requests.post(url=self.url, json=data).json()
        return response.get('result', [])

    def get_available_host(self):
        data = {
            "jsonrpc": "2.0",
            "method": "host.get",
            "params": {
                "filter": {
                    "available": 1
                },
                "output": [
                    "hostid",
                    "host"
                ],
                "selectInterfaces": [
                    "interfaceid",
                    "ip"
                ]
            },
            "id": 1,
            "auth": self.auth
        }

        response = requests.post(url=self.url, json=data).json()
        return response.get('result', [])

    def delete_host(self, host_id):
        data = {
            "jsonrpc": "2.0",
            "method": "host.delete",
            "params": [
                host_id
            ],
            "auth": self.auth,
            "id": 1
        }

        response = requests.post(url=self.url, json=data).json()
        return response.get('result', {})

    def make_host_available(self, host_id):
        data = {
            "jsonrpc": "2.0",
            "method": "host.update",
            "params": {
                "hostid": host_id,
                "available": 1
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        return response.get('result', {})

    def make_host_no_available(self, host_id):
        data = {
            "jsonrpc": "2.0",
            "method": "host.update",
            "params": {
                "hostid": host_id,
                "available": 2
            },
            "auth": self.auth,
            "id": 1
        }
        response = requests.post(url=self.url, json=data).json()
        return response.get('result', {})

# if __name__ == '__main__':
#     url = 'http://192.168.8.129/zabbix/api_jsonrpc.php'
#     za = Zabbix(url, 'Admin', 'zabbix')
#     # za = Zabbix('dragonstack', 'zabbix')
#     # print(za.historys_get('Pi9', "system.cpu.util[,user]", 0, 10))
#     print(za._historys_get("29654", 0, 1))
